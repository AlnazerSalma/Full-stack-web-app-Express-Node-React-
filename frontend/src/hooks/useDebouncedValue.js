// useDebouncedValue.js
import { useState, useEffect } from "react";

const useDebouncedValue = (value, delay = 200) => {
    const [debouncedValue, setDebouncedValue] = useState(value);

// Update query as soon as searchInput changes
useEffect(() => {
    // (Debounce) wait 200ms after user stop typing before 
    const handler = setTimeout(() => setDebouncedValue(value), delay);

    return () => clearTimeout(handler);
  }, [value, delay]);

  return debouncedValue;
};

export default useDebouncedValue;
