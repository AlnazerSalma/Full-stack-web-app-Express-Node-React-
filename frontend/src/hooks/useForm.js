import { useCallback } from "react";
import * as Yup from "yup";
import usePostFormData from "../utils/usePostFormData";

const useForm = (url) => {
  const { sendRequest, isLoading, error, data } = usePostFormData();

  const initialValues = {
    user_name: "",
    user_email: "",
    message: "",
  };

  const validationSchema = Yup.object({
    user_name: Yup.string().required("Name is required"),
    user_email: Yup.string()
      .email("Invalid email address")
      .required("Email is required"),
    message: Yup.string().required("Message is required"),
  });

  const onSubmit = useCallback(
    async (values, { resetForm, setSubmitting }) => {
      try {
        const res = await sendRequest(url, values);
        console.log("Response from backend:", res);
        resetForm();
      } catch (err) {
        console.error("Error sending to backend:", err);
      } finally {
        setSubmitting(false);
      }
    },
    [sendRequest, url]
  );

  return {
    initialValues,
    validationSchema,
    onSubmit,
    isLoading,
    error,
    data,
  };
};

export default useForm;
