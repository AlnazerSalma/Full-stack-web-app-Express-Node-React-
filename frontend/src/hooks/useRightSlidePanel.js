import { useState, useCallback } from "react";
function useRightSlidePanel() {
  const [selectedItem, setSelectedItem] = useState(null);
  const [isOpen, setIsOpen] = useState(false);

  const openPanel = useCallback((item) => {
    setSelectedItem(item);
    setIsOpen(true);
  }, []);

  const closePanel = useCallback(() => {
    setIsOpen(false);
    setSelectedItem(null);
  }, []);

  return {
    isOpen,
    selectedItem,
    openPanel,
    closePanel,
  };
}

export default useRightSlidePanel;
