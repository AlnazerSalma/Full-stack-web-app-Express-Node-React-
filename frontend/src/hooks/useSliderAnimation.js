import { useEffect } from "react";

const useSliderAnimation = (ref, items, duration = 40) => {
  useEffect(() => {
    if (items && ref.current) {
      ref.current.style.animation = `${duration}s linear infinite slideAnimation`;
    }
  }, [items, ref, duration]);
};

export default useSliderAnimation;
