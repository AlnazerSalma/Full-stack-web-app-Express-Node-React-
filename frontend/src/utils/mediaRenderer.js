import React from "react";

const mediaRenderer = ({ src, type, alt = "", className = "" }) => {
  if (!src) return null;

  if (type === "video" || src.endsWith(".mp4")) {
    return (
      <video
        src={src}
        autoPlay
        loop
        muted
        playsInline
        className={className}
      />
    );
  }

  return <img src={src} alt={alt} className={className} />;
};

export default mediaRenderer;
