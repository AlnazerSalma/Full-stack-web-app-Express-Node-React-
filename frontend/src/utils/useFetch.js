import { useEffect, useState } from "react";
import axios from "axios";

const useFetch = (url, maxRetries = 3, retryDelay = 1000) => {
  const [data, setData] = useState([]);
  const [loading, setLoading] = useState("Loading...");
  const [error, setError] = useState(null);

  useEffect(() => {
    let retries = 0;
    let isCancelled = false;

    const fetchData = async () => {
      setLoading("Loading...");
      while (retries < maxRetries) {
        try {
          const response = await axios.get(url);
          if (!isCancelled) {
            setData(response.data);
            setError(null);
            setLoading(null);
          }
          return;
        } catch (err) {
          retries += 1;
          if (retries >= maxRetries && !isCancelled) {
            setError("Failed to fetch data. Please try again later.");
            setLoading(null);
          } else {
            await new Promise((res) => setTimeout(res, retryDelay));
          }
        }
      }
    };

    fetchData();

    return () => {
      isCancelled = true;
    };
  }, [url, maxRetries, retryDelay]);

  return { data, loading, error };
};

export default useFetch;
