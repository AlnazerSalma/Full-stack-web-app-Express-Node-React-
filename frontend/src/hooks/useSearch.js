import { useState} from "react";
import useFetch from "../utils/useFetch"; 
import useDebouncedValue from "./useDebouncedValue";

const useSearch = (baseUrl, initialQuery = "", searchParam = "q") => {
    const [searchInput, setSearchInput] = useState(initialQuery);
    const debouncedQuery = useDebouncedValue(searchInput, 200);

    const url = debouncedQuery
      ? `${baseUrl}/search?${searchParam}=${encodeURIComponent(debouncedQuery)}`
      : baseUrl;

    const { data, loading, error } = useFetch(url);

    return {
    data,
    loading,
    error,
    searchInput,
    setSearchInput,
  };
};

export default useSearch;
