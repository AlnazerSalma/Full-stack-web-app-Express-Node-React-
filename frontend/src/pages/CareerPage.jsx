import React, { lazy, Suspense } from "react";
import { Container, Row, Col } from "react-bootstrap";
import CareerOpenPosition from "../components/career/CareerOpenPosition";
import useFetch from "../utils/useFetch";
import useRightSlidePanel from "../hooks/useRightSlidePanel";
import "../style/Careers.css";

const CareerRightSlidePanel = lazy(() =>
  import("../components/right_slider_panel/CareerRightSlidePanel")
);

function CareerPage() {
  const {
    data: careerData,
    loading,
    error,
  } = useFetch("http://localhost:5000/api/careers");

  const {
    isOpen,
    selectedItem,
    openPanel,
    closePanel,
  } = useRightSlidePanel();

  return (
    <Container fluid className="careers-section">
      <Container>
        <Row>
          <Col className="text-start">
            <h1 className="careers-title">Working at Haptic</h1>
            <h1 className="careers-subtitle">
              Great work starts with <br /> great people —{" "}
              <strong className="purple">join us.</strong>
            </h1>
            {loading && <p className="loading-message">{loading}</p>}
              {error && <p className="error-message">{error}</p>}
            {!loading && !error && (
              <>
                <h1 className="careers-open-positions">⚪ Open position</h1>
                {careerData.map((position, idx) => (
                  <CareerOpenPosition
                    key={idx}
                    title={position.title}
                    location={position.location}
                    type={position.type}
                    onClick={() => openPanel(position)}
                  />
                ))}
              </>
            )}
          </Col>
        </Row>
      </Container>

      {selectedItem && (
        <Suspense fallback={<div>Loading details...</div>}>
        <CareerRightSlidePanel
           isOpen={isOpen}
           onClose={closePanel}
           position={selectedItem}
        />
        </Suspense>
      )}
    </Container>
  );
}

export default CareerPage;
