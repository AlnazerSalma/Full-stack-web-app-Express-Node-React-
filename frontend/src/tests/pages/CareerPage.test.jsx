import React, { Suspense } from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import CareerPage from "../../pages/CareerPage";
import * as fetchHook from "../../utils/useFetch";
import * as panelHook from "../../hooks/useRightSlidePanel";
import mockOpenposition from "../__mocks_data__/mockOpenposition";

jest.mock("../../components/right_slider_panel/CareerRightSlidePanel.jsx", () => () => (
  <div data-testid="right-slide-panel">Right Slide Panel</div>
));

jest.mock("../../components/career/CareerOpenPosition", () => ({ title, onClick }) => (
  <div data-testid="role" onClick={onClick}>
    {title}
  </div>
));


describe("CareerPage", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("shows loading state", () => {
    jest.spyOn(fetchHook, "default").mockReturnValue({
      data: [],
      loading: "Loading...",
      error: null,
    });

    jest.spyOn(panelHook, "default").mockReturnValue({
      isOpen: false,
      selectedItem: null,
      openPanel: jest.fn(),
      closePanel: jest.fn(),
    });

    render(<CareerPage />);
    expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
  });

  it("shows error state", () => {
    jest.spyOn(fetchHook, "default").mockReturnValue({
      data: [],
      loading: false,
      error: "Failed to fetch",
    });

    jest.spyOn(panelHook, "default").mockReturnValue({
      isOpen: false,
      selectedItem: null,
      openPanel: jest.fn(),
      closePanel: jest.fn(),
    });

    render(<CareerPage />);
    expect(screen.getByText(/Failed to fetch/i)).toBeInTheDocument();
  });

  it("renders job roles and opens the right panel when clicked",() => {
 
    const openPanel = jest.fn();

    jest.spyOn(fetchHook, "default").mockReturnValue({
      data: mockOpenposition,
      loading: false,
      error: null,
    });

    jest.spyOn(panelHook, "default").mockReturnValue({
      isOpen: false,
      selectedItem: null,
      openPanel,
      closePanel: jest.fn(),
    });

    render(<CareerPage />);

    const role = screen.getByTestId("role");
    expect(role).toBeInTheDocument();

    fireEvent.click(role);
    expect(openPanel).toHaveBeenCalledWith(mockOpenposition[0]);
  });

it("shows the right panel when a role is selected", async () => { 

  jest.spyOn(fetchHook, "default").mockReturnValue({
    data: mockOpenposition,
    loading: false,
    error: null,
  });

  jest.spyOn(panelHook, "default").mockReturnValue({
    isOpen: true,
    selectedItem: mockOpenposition[0],
    openPanel: jest.fn(),
    closePanel: jest.fn(),
  });

  render(
    <Suspense fallback={<div>Loading...</div>}>
      <CareerPage />
    </Suspense>
  );

  await waitFor(() => {
    expect(screen.getByTestId("right-slide-panel")).toBeInTheDocument();
  });
});
it("renders correctly and shows the right panel when a role is selected (snapshot test)", () => {

  jest.spyOn(fetchHook, "default").mockReturnValue({
    data: mockOpenposition,
    loading: false,
    error: null,
  });

  jest.spyOn(panelHook, "default").mockReturnValue({
    isOpen: true,
    selectedItem: mockOpenposition[0],
    openPanel: jest.fn(),
    closePanel: jest.fn(),
  });

  const { asFragment } = render(
    <Suspense fallback={<div>Loading...</div>}>
      <CareerPage />
    </Suspense>
  );

  expect(asFragment()).toMatchSnapshot();
});

});
