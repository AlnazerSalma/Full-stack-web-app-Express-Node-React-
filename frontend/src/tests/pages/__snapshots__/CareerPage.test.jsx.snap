// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CareerPage renders correctly and shows the right panel when a role is selected (snapshot test) 1`] = `
<DocumentFragment>
  <div
    class="careers-section container-fluid"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="text-start col"
        >
          <h1
            class="careers-title"
          >
            Working at Haptic
          </h1>
          <h1
            class="careers-subtitle"
          >
            Great work starts with 
            <br />
             great people — 
            <strong
              class="purple"
            >
              join us.
            </strong>
          </h1>
          <h1
            class="careers-open-positions"
          >
            ⚪ Open position
          </h1>
          <div
            data-testid="role"
          >
            Software Engineer
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="right-slide-panel"
    >
      Right Slide Panel
    </div>
  </div>
</DocumentFragment>
`;
